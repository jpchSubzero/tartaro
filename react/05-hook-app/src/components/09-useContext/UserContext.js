// Importamos la función de creación de contextos
import { createContext } from 'react';

// 1. Creamos y exportamos un contexto con valor null. Este se utilizará en el componente de más alto nivel de donde se necesite. Es decir, si existen 10 nivles de componentes (abuelo, padre, hijo, nieto, bisnieto, ..., 10mo) y el contexto se necesita en los 4 últimos niveles se deberá utilizar en el nivel 6 que contiene el 7, 8, 9 y 10. En este caso irá en el MainApp ya que este contiene a todos
export const UserContext = createContext(null);