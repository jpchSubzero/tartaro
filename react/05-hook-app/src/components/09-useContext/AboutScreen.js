import React, { useContext } from 'react';
import { UserContext } from './UserContext';


export const AboutScreen = () => {

    // 1. Como dejamos de enviar solo el valor y ahora es el state, destructuramos el objeto para tener user y setUser. En este caso solo user porque no se utiliza setUser
    const { user, setUser } = useContext(UserContext);

    const handleClick = () => {
        setUser({});
    };

    return (
        <div>
            <h1>AboutScreen</h1>
            <hr />

            {/* 3. Agregamos un bot√≥n para borrar el user */}
            <button className="btn btn-danger" onClick={ handleClick }>Log Out</button>

            {/* 2. Agregar el user como json */}
            <pre className="container">{ JSON.stringify(user, null, 3) }</pre>

        </div>    
    )
}
