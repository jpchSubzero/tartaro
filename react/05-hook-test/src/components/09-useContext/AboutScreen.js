import React, { useContext } from 'react';
import { UserContext } from './UserContext';


export const AboutScreen = () => {
    const { user, setUser } = useContext(UserContext);

    const handleClick = () => {
        setUser({});
    };

    return (
        <div>
            <h1>AboutScreen</h1>
            <hr />

            <button className="btn btn-danger" onClick={ handleClick }>Log Out</button>

            <pre className="container">{ JSON.stringify(user, null, 3) }</pre>

        </div>    
    )
}
