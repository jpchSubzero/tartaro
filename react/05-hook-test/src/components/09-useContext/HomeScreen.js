import React, { useContext } from 'react';
import { UserContext } from './UserContext';

export const HomeScreen = () => {

    // 1. Agregarmos la referencia al contexto con el hook useContext. Devuelve el valor asignado en el componente que se defini√≥ para almacenar, en ese caso UserContext
    // const userContext = useContext(UserContext);
    
    // 2. Como dejamos de enviar solo el valor y ahora es el state, destructuramos el objeto para tener user y setUser. En este caso solo user porque no se utiliza setUser
    const { user } = useContext(UserContext);

    console.log(user);

    return (
        <div>
            <h1>HomeScreen</h1>
            <hr />

            {/* 3. Agregarmos un <pre> para ver los valores del user. La clase container es para que se vea mejor */}
            <pre className="container">{ JSON.stringify(user, null, 3) }</pre>
        </div>
    )
}
