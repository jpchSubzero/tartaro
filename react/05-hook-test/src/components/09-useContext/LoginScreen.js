import React, { useContext } from 'react';
import { UserContext } from './UserContext';

export const LoginScreen = () => {
    const { setUser } = useContext(UserContext);

    const newUser = {
        id: 1234,
        name: 'Nombre del usuario',
        email: 'usuario@usuario.com'
    };

    return (
        <div>
            <h1>LoginScreen</h1>
            <hr />

            <button className="btn btn-primary" onClick={() => setUser(newUser)}>Log In</button>
        </div>    
    )
}

