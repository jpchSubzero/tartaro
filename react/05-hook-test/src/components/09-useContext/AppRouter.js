import React from 'react';
import { AboutScreen } from './AboutScreen';
import { LoginScreen } from './LoginScreen';
import { HomeScreen } from './HomeScreen';

// Importar componentes para el uso de router. Se puede copiar directamente de la documentación.
// https://reactrouter.com/web/guides/quick-start
// as Router es para crear un alias y no utilizar todo el nombre del componente BrowserRouter.
// Para evitar los warnings de Link es mejor quitarlo porque se lo va a utilizar en el navbar
import { BrowserRouter as Router, Switch, Route, Redirect } from "react-router-dom";
import { NavBar } from './NavBar';

export const AppRouter = () => {
    return (
        // Especifica que es un router
        <Router>
            {/* Contenedor de todos los componentes que se vayan a utilizar. Es opcional */}
            <div>
                {/* 6. Agregar el NavBar después de probar las rutas manualmente */}
                <NavBar />

                {/* 7. Agregar un div container para que mejore la presentación */}
                <div className="container mt-5">
                    {/* Definir las rutas a utilizar */}
                    <Switch>
                        {/* 1. Es necesario el / en la ruta */}
                        {/* <Route path="/about" component={ AboutScreen } />
                        <Route path="/login" component={ LoginScreen } /> */}

                        {/* 2. Si lo ponemos primero siempre se va a mostrar el home porque valida por coincidencia parcial y si el path tiene / se iria al home. Para solucionar se puede ordenar los Route del más específico al más general (/user/create, /user/update, /about, /login, /user, /) o agregar el atributo exact a Route para que haga una evaluación completa de la ruta */}
                        {/* <Route path="/" component={ HomeScreen } /> */}

                        <Route exact path="/" component={ HomeScreen } />
                        <Route exact path="/about" component={ AboutScreen } />
                        <Route exact path="/login" component={ LoginScreen } />

                        {/* 3. Ruta por defecto para manejar rutas que no existen y deben llevar a una por defecto, por lo general a la página de Not Found. No cambia el path del navegador */}
                        {/* Comparando con cualquiera que no exista */}
                        {/* <Route exact path='*' component={HomeScreen} /> */}

                        {/* 4. Simplemente enviar a la ruta por defecto si no encuentra las anteriores */}
                        {/* <Route component={HomeScreen} /> */}

                        {/* 5. Esta opción redirecciona a una ruta por defecto. Implica refresh pero tambien cambia el path del navegador */}
                        <Redirect to="/" />
                    </Switch>
                </div>
            </div>
        </Router>
    )
}
