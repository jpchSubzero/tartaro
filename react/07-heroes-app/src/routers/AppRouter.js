// Desde https://reactrouter.com/web/guides/quick-start copiar los imports y <Router>

import React from "react";
import {
  BrowserRouter as Router,
  Switch,
  Route,
} from "react-router-dom";
import LoginScreen from "../components/login/LoginScreen";
// import MarvelScreen from "../components/marvel/MarvelScreen";
// import { Navbar } from "../components/ui/NavBar";
import DaskboardRoutes from "./DaskboardRoutes";

const AppRouter = () => {
    return (
        <Router>
        <div>
            {/* Reemplazar el <nav> del ejemplo por el NavBar que creamos */}
            {/* Quitamos el navbar para que no se presente en todas las p√°ginas ya que eso hace que salga en el login */}
            {/* <Navbar /> */}
            <Switch>

                {/* Reemplazar las rutas por las que vamos utilizar */}
                <Route exact path="/login" component={ LoginScreen } />

                {/* <Route exact path="/" component={ MarvelScreen } /> */}

                {/* Cambiamos el componente MarvelScreen por el component que tiene las rutas una vez logueado */}
                {/* Quitamos el exact porque la idea es que si no es login nos lleve a las otras rutas de Daskboard */}
                <Route path="/" component={ DaskboardRoutes } />
            </Switch>
        </div>
      </Router>
    )
}

export default AppRouter;
