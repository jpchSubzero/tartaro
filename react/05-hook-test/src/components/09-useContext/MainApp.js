import React, { useState } from 'react';
import { AppRouter } from './AppRouter';
import { UserContext } from './UserContext';

export const MainApp = () => {

    // 3. Agregamos el objeto/información a compartir
    // const user = {
    //     id: 1234,
    //     name: 'Nombre del usuario',
    //     email: 'usuario@usuario.com'
    // }

    // 5. Reemplazamos el objeto por el state. COmo estabamos enviando user y el state tiene el mismo nombre app va a seguir igual solo que con un objeto vacío
    const [user, setUser] = useState({});

    return (
        // <div>
        //     <h1>MainApp</h1>
        //     <hr />
        //     <AppRouter />
        // </div>    

        // 1. Podemos dejar solo el AppRouter ya que es quien renderizará lo que se vaya pidiendo
        // <AppRouter />

        // 2. Utilizar el contexto creado para manejar la comunicación entre componentes que estan al mismo nivel. Al revisar en la consola de react se ve una estructura mucho más compleja que antes
        // <UserContext.Provider>
        //     <AppRouter />
        // </UserContext.Provider>

        // 4. Enviamos la información a través del atributo value. Al reviasr en consola se ve que ya tiene la prop el contexto
        // <UserContext.Provider value={ user }>

        // 6. Enviamos tanto el valor como la función en un objeto nuevo, { user, setUser }. Como usamos los mismo nombres se omite : { user:user, setUser:setUser }
        <UserContext.Provider value={ { user, setUser } }>
            <AppRouter />
        </UserContext.Provider>
    )
}
