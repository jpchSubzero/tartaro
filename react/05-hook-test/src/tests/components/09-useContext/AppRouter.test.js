import React from 'react';
import { mount } from 'enzyme';

import { AppRouter } from '../../../components/09-useContext/AppRouter';
import { UserContext } from '../../../components/09-useContext/UserContext';

describe('Pruebas con AppRouter', () => {
    // const wrapper = mount(<AppRouter />);

    // Agregar el user que necesitan los componentes
    const user = {
        id: 1234,
        name: 'Nombre del usuario',
        email: 'usuario@usuario.com'
    };

    // Agregar el provider para acceder al contexto
    const wrapper = mount(
        <UserContext.Provider value={ { user } }>
            <AppRouter />
        </UserContext.Provider>
    );

    test('Debe mostrarse correctamente', () => {
        // Al probar directamente con mount y snapshot nos da un error porque varios componentes usan user desde el contexto (HomeScreen, AboutScreen y LoginScreen)
        expect(wrapper).toMatchSnapshot();
    });
    
    // Otra puede que se podría realizar es validar rutas con parámetros pero aún no se está en condiciones
    // No es necesario validar ni switch ni route porque son componentes de react y tuvieron sus pruebas respectivas
});
