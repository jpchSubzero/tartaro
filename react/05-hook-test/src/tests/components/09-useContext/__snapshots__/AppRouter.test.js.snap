// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas con AppRouter Debe mostrarse correctamente 1`] = `
<div>
  <nav
    className="navbar navbar-expand-sm navbar-dark bg-dark"
  >
    <div
      className="container-fluid"
    >
      <a
        aria-current="page"
        className="navbar-brand active"
        href="/"
        onClick={[Function]}
        style={Object {}}
      >
        useContextApp
      </a>
      <button
        aria-controls="navbarNavAltMarkup"
        aria-expanded="false"
        aria-label="Toggle navigation"
        className="navbar-toggler"
        data-bs-target="#navbarNavAltMarkup"
        data-bs-toggle="collapse"
        type="button"
      >
        <span
          className="navbar-toggler-icon"
        />
      </button>
      <div
        className="collapse navbar-collapse"
        id="navbarNavAltMarkup"
      >
        <div
          className="navbar-nav"
        >
          <a
            aria-current="page"
            className="nav-link active"
            href="/"
            onClick={[Function]}
            style={Object {}}
          >
            Home
          </a>
          <a
            aria-current={null}
            className="nav-link"
            href="/login"
            onClick={[Function]}
          >
            Login
          </a>
          <a
            aria-current={null}
            className="nav-link"
            href="/about"
            onClick={[Function]}
          >
            About
          </a>
        </div>
      </div>
    </div>
  </nav>
  <div
    className="container mt-5"
  >
    <div>
      <h1>
        HomeScreen
      </h1>
      <hr />
      <pre
        className="container"
      >
        {
   "id": 1234,
   "name": "Nombre del usuario",
   "email": "usuario@usuario.com"
}
      </pre>
    </div>
  </div>
</div>
`;
